### PR TITLE
fix --DRT-gcopt string argument parsing

### PR DESCRIPTION
**Pull request by MartinNowak**
- would match more than just an identifier, e.g.
  'gc:conservative profile:1' was parsed as gc='conservative profile:1'
- was restricted to known GC names, instead we now check
  that one GC was initialized
- parse const(char)[] arguments as simple identifier [_a-zA-Z][_a-zA-Z0-9]*

### CODE REVIEW

File: src/gc/config.d [hunk 1]
```diff
@@ -26,7 +26,7 @@ struct Config
 {
     bool disable;            // start disabled
     ubyte profile;           // enable profiling with summary when terminating program
-    string gc = "conservative"; // select gc implementation conservative|manual
+    const(char)[] gc = "conservative"; // select gc implementation conservative|manual
```
Comment (WalterBright) [inline src/gc/config.d]: why this change?

Comment (MartinNowak) [inline src/gc/config.d]: If we don't know all GCs that can be selected (b/c one w/ a new name can be linked in), then we can't have all names as string symbols. I'm now just taking a slice of the DRT-gcopt argument, which can come from the command line, a string constant in the binary, or if enabled from the environment (see [gc_config](https://dlang.org/spec/garbage.html#gc_config)).
We could malloc and copy the string to make it immutable, but at the time gc_init runs, no usercode has run, so it's hardly possible to change cmdline or env variables. And if someone does it, they simply have to deal w/ this.

**Comment by MartinNowak** [comment]
Ping, also @Jebbs who wrote the former code.

File: src/gc/config.d [hunk 2]
```diff
@@ -190,27 +190,10 @@ bool parse(T:const(char)[])(const(char)[] optname, ref const(char)[] str, ref T
...
-    }
-    else
-    {
-        return parseError("conservative or manual", optname, str);
-    }
-
+    // [_a-zA-Z][_a-zA-Z0-9]*]
+    if (str[0] != '_' && !isalpha(str[0]))
+        return parseError("an identifier", optname, str);
+    res = str[0 .. $ - str[1 .. $].find!(c => c != '_' && !isalnum(c)).length];
```
Comment (rainers) [inline src/gc/config.d]: We are pretty permissive about characters in option names, we should not be more picky with the name of the GC. I think everything but the same terminators should be accepted (i.e. ' ', ':' and '='). This avoids having to camel-case names composed of multiple words, e.g. `mark-and-sweep`.

File: src/gc/config.d (hunk 1 above)
Comment (rainers) [inline src/gc/config.d]: `parse` in `gc.config.initialize` receives a string, doesn't that allow to keep the options as strings aswell? It would be surprising if the referred string would be changed after initialization...

**Comment by MartinNowak** [comment]
Both done

**Comment by rainers** [comment]
Auto-merge toggled on
